pipelines/notion/client.py

```diff
@@ -0,0 +1,156 @@
+        subresource: Optional[str] = None,
+        query_params: Optional[Dict[str, Any]] = None,
+        payload: Optional[Dict[str, Any]] = None,
+    ) -> Dict[str, Any]:
+        """Sends a payload to the Notion API using the POST method.
+
+        Args:
+            resource (str): The resource to send the payload to.
+            resource_id (str): The id of the resource.
+            subresource (str, optional): The subresource to send the payload to.
+                Defaults to None.
+            query_params (Dict[str, Any], optional): The query parameters to send
+                with the payload. Defaults to None.
+            payload (Dict[str, Any], optional): The payload to send. Defaults to None.
+
+        Returns:
+            Dict[str, Any]: The response from the Notion API.
+
+        Raises:
+            requests.HTTPError: If the response from the Notion API is not 200.
+        """
+
+        url = self.get_endpoint(resource, resource_id, subresource)
+        headers = self._create_headers()
+
+        if payload is None:
+            payload = {}
+
+        filtered_payload = self._filter_out_none_values(payload)
+
+        response = requests.post(
+            url, headers=headers, params=query_params, json=filtered_payload
+        )
+        response.raise_for_status()
+        return response.json()
+
+    def search(
+        self,
+        query: Optional[str] = None,
+        filter_criteria: Optional[Dict[str, Any]] = None,
+        sort: Optional[Dict[str, Any]] = None,
+        start_cursor: Optional[str] = None,
+        page_size: Optional[int] = None,
+    ) -> Dict[str, Any]:
+        """Searches all parent or child pages and databases that have been
+        shared with an integration.
+
+        Notion API Reference. Search:
+            https://developers.notion.com/reference/post-search
+
+        Args:
+            query (str, optional): The string to search for. Defaults to None.
+            filter_criteria (Dict[str, Any], optional): The filter to apply to
+                the results.
+            sort (Dict[str, Any], optional): The sort to apply to the results.
+            start_cursor (str, optional): The cursor to start the query at.
+                Defaults to None.
+            page_size (int, optional): The number of results to return.
+                Defaults to None.
+
+        Yields:
+            Dict[str, Any]: A result from the search.
+        """
+        has_more = True
+
+        while has_more:
+            payload = {
+                'query': query,
+                'sort': sort,
+                'filter': filter_criteria,
+                'start_cursor': start_cursor,
+                'page_size': page_size,
+            }
+
+            filtered_payload = self._filter_out_none_values(payload)
+
+            response = self.send_payload(
+                'search', '', payload=filtered_payload
+            )
+
+            for result in response.get('results', []):
```

Comment:
it is way faster to yield lists, not item by item. `dlt` will handle both lists and items same way, transparently for the user. list goes through pipeline as single item. if you use transformers, it is really way more efficient

Reply:
I use `search()` in `notion_databases()` to filter database ids myseld here:

```python
if database_ids is None:
    search_results = notion_client.search(
        filter_criteria={'value': 'database', 'property': 'object'}
    )
    database_ids = [
        {'id': result['id'], 'use_name': result['title'][0]['plain_text']}
        for result in search_results
    ]
```

If I yield a list, the usability of `search()` could suffer, since the calling code would need to handle chaining. For this particular case I'm not sure if the performance gains would be that visible. Happy to adjust if I'm wrong.

Reply:
yes you are right (I feel like GPT-4 now ;p) you are using search not to query data, just the database ids... then all is good and we should keep yielding single elements